File: c/algorithms/reverse.c
#include "reverse.h"

#include <stddef.h>
// #include <string.h>

static size_t strlen(const char* string) {
  size_t length = 0;
  while (string[length] != '\0') {
    ++length;
  }
  return length;
}

static void swap(char* i, char* j) {
  char temp = *i;
  *i = *j;
  *j = temp;
}

void reverse(char* string) {
  size_t length = strlen(string);
  if (length == 0) {
    return;
  }
  size_t low = 0;
  size_t high = length - 1;
  while (low < high) {
    swap(string + low, string + high);
    ++low;
    --high;
  }
}

File: python/algorithms/reverse.py
#!/usr/bin/env python


def reverse(iterable):
  """Reverses the iterable.

  Iterate through the collection with two pointers, one increasing from the
  beginning of the collection and the other decreasing from the end of the
  collection, and swap the items at the respective pointers. Iterations through
  the collection are repeated until the pointer that is increasing is no longer
  less than the pointer that is decreasing.

  * O(n) time complexity
  * O(1) space complexity

  Args:
    iterable: A collection that is iterable.
  Raises:
    TypeError: If iterable is not iterable.
  """
  try:
    _ = iter(iterable)
  except TypeError:
    raise TypeError('\'{}\' object is not iterable'.format(
        type(iterable).__name__))

  low = 0
  high = len(iterable) - 1
  while low < high:
    iterable[low], iterable[high] = iterable[high], iterable[low]
    low += 1
    high -= 1
